Feature anyof

CD3 opened the pull request:
Adds an 'anyof' rule that will allow a list of definitions to be given. Currently, all rules in a definition are checked, so a rule is a set of constraints that are AND'ed together. The 'anyof' key allows constraints to be OR'ed. This should work transparently will all existing rules and any user defined validators.

Simple example, taken from unit test
```
# prop1 must be either a number between 0 and 10
schema = {'prop1':  {'min': 0, 'max': 10}}
doc = {'prop1': 5}

self.assertSuccess(doc, schema)

# prop1 must be either a number between 0 and 10 or 100 and 110
schema = {'prop1':
                  {'anyof':
                   [{'min': 0, 'max': 10}, {'min': 100, 'max': 110}]}}
doc = {'prop1': 105}

self.assertSuccess(doc, schema)

# prop1 must be either a number between 0 and 10 or 100 and 110
schema = {'prop1':
                  {'anyof':
                   [{'min': 0, 'max': 10}, {'min': 100, 'max': 110}]}}
doc = {'prop1': 50}

self.assertValidationError(doc, schema)

```

This also satisfies #108. See the [unit test](https://github.com/CD3/cerberus/commit/53c6b72b90bf4827038252a29fc8be4f11a1a938) for the syntax.

funkyfuture commented on cerberus/cerberus.py:
```diff
@@ -643,6 +646,25 @@ def _validate_validator(self, validator, field, value):
+            definition = anyof[i]
+            self._errors = {}
+            self._validate_definition(definition, field, value)
+            errorstack["definition %d" % i] = self._errors.get(field, {})
```
shouldn't that be after an `else`-statement after the next `if`!?

funkyfuture commented on cerberus/cerberus.py:
```diff
@@ -643,6 +646,25 @@ def _validate_validator(self, validator, field, value):
         validator(field, value, self._error)
 
+    def _validate_anyof(self, anyof, field, value):
+        if isinstance(anyof, Mapping):
```
is that to allow this?:
```python
schema = {'property': {'anyof': {one_rule}}}
```

CD3 replied:
Yes

funkyfuture commented:
looks good, but the problem is that you can't have multiple `anyof` in one schema. which you would need if you wanted to `and` two `anyof`-sets. but i can't come up with a pragmatic use-case and it seems obvious how a `allof` would be implemented.

i would undo any changes concerning the usage of `type`-rule with a list of valid types. does anyone disagree?

you may consider to squash your commits into one.

CD3 commented:
hmm, good point. I'm not sure logical OR and AND commute or not.

I'll squash the commits.

What do suggest. Should I implement an `allof` an create a new pull requrest or add it to this one?

I think that the `type`  revert should be handled separably, it will probably require existing unit tests to be modified. Although, the current syntax may be a useful shorthand since wanting a list of types is clearly a common thing.

funkyfuture commented:
> What do suggest. Should I implement an allof an create a new pull requrest or add it to this one?

i'm not sure if this is actually needed in the real world.

also, i'm still thinking about coercion in that context. not sure, if it's a good idea to allow this.

what i meant: as i'm the one who did the changes regarding `type`, i would also undo them. but maybe there are objections.

btw, it *could* (didn't test it, just played around) be done in three lines less code:
```python
errorstack = {}

for definition, i in anyof, len(anyof):
    v = self.__get_child_validator(schema={field: definition})
    if v.validate({field}):
        valid = True
        break
    else:
        errorstack["definition %d" % i] = v.errors
    
if errorstack:
   self._error(field, errorstack)
```

CD3 commented:
well, if that works it would not require the `validate_definition` function, which would be nice. let me try it out.

funkyfuture commented:
don't forget inline- and end-user-documentation. :-D

CD3 commented:
of course, I'm new.

funkyfuture commented:
`Validator.validate_schmema` should ensure that only `Sequence`s of `Mapping`s are used.

CD3 commented:
Well, using a child validator complicates things. The child validator runs the schema through `validate_schema` which fails on my unit test.

funkyfuture commented:
what's failing exactly?

CD3 commented:
It doesn't like that a `schema` rule exists without a `type` key specifying `dict` or `list`. So it wants to see

```
{ 'anyof' : [ {'type' : 'dict', 'schema' : {...}}, {'type' : 'dict', 'schema' : {...}}]
```

instead of
```
{ 'type' : 'dict', 'anyof' : [ {'schema' : {...}}, {'schema' : {...}}]}
```

I didn't look at it anymore after I figured out that was what was wrong. The `_validate_anyof` function would need access to the level above its argument, which would require a workaround.

I'd actually like to change the `validate_schema` function to allow this,
```
{ 'type' : 'dict', 'schema' : { 'anyof' : [ {...}, {...}]}}
```
right now it complains that entries in 'schema' cannot be lists.

nicolaiarocci commented:
Hey guys, just wanted to let you know that I'm following the discussion and I like what you are doing here. 

> right now it complains that entries in 'schema' cannot be lists.

You could amend `validate_schema` to allow for `anyof`, it would not break backward compatibility I think.

CD3 commented:
Thanks @nicolaiarocci, I actually wondered if I should stop looking at this until I heard from you.

A couple of things.

I added a `allof` constraint that does the same thing as `anyof`. Who knows, it may be useful and it seems more symmetric to me.

I also updated `validate_schema` to handle `anyof` and `allof` constraints. Basically, if one of these constraints is run into, it check that the schema formed by replacing the `*of` constraint with each of its values is valid. For example, if you have

```
schema = {'doc' : {'type' : 'number', 'anyof' : [ {'max' : 0}, {'min' : 10} ] } }
```
 then `validate_schema` would check that both
```
schema = {'doc' : {'type' : 'number', 'max' : 0 } }
```
and
```
schema = {'doc' : {'type' : 'number', 'max' : 0 } }
```
are valid schemas.

The addition to `validate_schema` failed the Trivis CI, I used dict comprehension which is not supported in 2.6.

@funkyfuture, I don't think you should remove list support from type constraints yet, `anyof` does not quite duplicate its behavior. If you wanted to specify multiple types, you have to do this

```
schema = {'parm' : {'anyof' : [ {'type : 'string'}, {'type' : 'number'} ] }
```

Its not possible to do

```
schema = {'parm' : {'type' : { 'anyof' : ['string', 'number'] } } }
```

Initially I thought it was because of the `validate_schema` function, but it is actually more than that. The function that validates `type` will be passed `{ 'anyof' : ['string', 'number'] } }`, which it does not know how to handle. The same is true for all other constraints too.

So, I think that adding support for prefixing keys with `anyof_` and `allof_` would also be useful (I think `anyof` constraint is still useful as is)

I'll fix the 2.6 compatibility and add user documentation if all agree that this feature will be accepted.

nicolaiarocci commented on cerberus/tests/tests.py:
```diff
@@ -850,6 +850,159 @@ def test_validated(self):
+
+        self.assertValidationError(doc, schema)
+
+        # prop1 must be either be less than 0 or less than 10
```
you probably mean "equal or greater than 0 or equal or greater than 10"

nicolaiarocci commented on cerberus/tests/tests.py:
```diff
@@ -850,6 +850,159 @@ def test_validated(self):
+        doc = {'prop1': 11}
+        self.assertFail(doc, schema)
+
+        # prop1 has to be a float and a string
```
"float and an integer"

nicolaiarocci commented on cerberus/tests/tests.py:
```diff
@@ -850,6 +850,159 @@ def test_validated(self):
+        doc = {'prop1': '5.5'}
+        self.assertFail(doc, schema)
+
+        doc = {'prop1': 5}
```
test missing?

nicolaiarocci commented:
Green light from me. Make sure docs are updated appropriately. Tests are likely to serve as valid documentation examples too.

funkyfuture commented:
as you did the `allof`, we could complete the set of logical operators with `oneoff` (*one*, not more, must match) and `noneoff`.

when i'm home in some hours, i'd like to investigate shortly on that issue with a child-validator.

funkyfuture commented on cerberus/cerberus.py:
```diff
@@ -371,6 +374,20 @@ def update_to_valueschema(schema, warning_printed=False):
                             errors.ERROR_SCHEMA_TYPE.format(field))
                 elif constraint in self.special_rules:
                     pass
+                elif constraint == 'anyof' or constraint == 'allof':
```
`elif constraint in ('anyof', 'allof'):` is shorter and better extendable.

funkyfuture commented:
> I don't think you should remove list support from type constraints yet, anyof does not quite duplicate its behavior.

i know, but i'd prefer one way to do it, and that would be the generic one.

please add a note to the documentation that usage of these rules is not tested in conjunction with coercion and not recommended.

funkyfuture commented:
fyi, as i would propably refactor to a usage of another instance as sketched above when i'm going to implement #93 anyway i pulled your code and tried it.

i also have that failing schema-validation on one of your tests (`test_anyof_schema`) as well. strangely when i follow the execution with a debugger, the code that will later be displayed as raising a `SchemaError` is actually not executed, and the `validate_schema`-method terminate normally.

CD3 commented:
I think it is good to go. I won't have time to add `noneof` or `oneof` for a couple of days, but I'll work on it.

nicolaiarocci commented:
Merged with https://github.com/nicolaiarocci/cerberus/commit/24f80286a4da33adff691e71ba3044939089aa26

Thanks!
